text/2021-12-09-logging.md

```diff
@@ -0,0 +1,302 @@
+
+## Motivation
+
+<!-- Why are we doing this? What use cases does it support? What is the expected
+outcome? -->
+
+Logging in each component in Chaos Mesh is bad. Related issue:
+https://github.com/chaos-mesh/chaos-mesh/issues/2149. Because of the mess on
+logging, the debugging and profiling of the components are difficult.
+
+This proposal aims to improve the logging observability of the components and
+the development experience on printing logs.
+
+After passed this proposal, I would write a developing guide for logging in the
+repo of Chaos Mesh.
+
+## Requirements
+
+- structured logging
+- leveled logging
+- runtime level configure
+- could logging in everywhere (every line of code)
+
+## Detailed design
+
+<!-- This is the bulk of the RFC. Explain the design in enough detail that:
+
+- It is reasonably clear how the feature would be implemented.
+- Corner cases are dissected by example.
+- How the feature is used. -->
+
+We are going to use [logr](https://github.com/go-logr/logr) and
+[zap](https://github.com/uber-go/zap) (with
+[zapr](https://github.com/go-logr/zapr) as the shim between them) to construct
+the logging facilities.
+
+[logr](https://github.com/go-logr/logr) is a library that provides a common
+interface for logging, in other words, logging facade. It is designed to be used
+by different logging frameworks(likes the SLF4J API in Java), with using logr,
+we could switch from one logging backend to another without many changes.
+
+[zap](https://github.com/uber-go/zap) is a high performance logging framework.
+[zapr](https://github.com/go-logr/zapr) is the logr implementation using zap.
+
+### Global Logger
+
+Here would exists a global logger at `pkg/log`, which marked as deprecated and
+only for the compatibility with the old code. Global Logger could be accessed by
+`log.L()`.
+
+The initial value of the global logger is `logr.DiscardLogger`, which means no
+log would be printed anywhere. The global logger should be replaced ONLY ONCE
+when the application starts.
+
+### Logger as dependency
+
+The suggested way to use the logger is to use it as a dependency, and make
+dependency explicit.
+
+GOOD Example:
+
+```go
+type SomeStruct struct {
+  state string
+  otherField string
+  logger logr.Logger
+}
+
+func NewSomeStruct(state string, otherField string, logger logr.Logger) *SomeStruct {
+  return &SomeStruct{
+    state: state,
+    otherField: otherField,
+    logger: logger,
+  }
+}
+
+func (it *SomeStruct) DoSomething()  {
+  it.logger.Info("Doing something")
+}
+```
+
+MUST NOT use the implicitly logger in the constructor.
+
+BAD example:
+
+```go
+func NewSomeStruct(state string, otherField string) *SomeStruct {
+  return &SomeStruct{
+    state: state,
+    otherField: otherField,
+    logger: somewhere.DefaultLogger(),
+  }
+}
+```
+
+And logger should be a parameter of the function, not a global variable.
+
+GOOD example:
+
+```go
+func doSomething(logger logr.Logger) {
```

Comment:
if choose this one, we need to add log as a parameter to the most functions

Reply:
Well, for existing codes, it could still use the global logger for avoiding too many changes.

But for incoming new codes, I think if it requires the logger, it's mostly complex enough to be refactored as "Objects". Or use some tricks like "functional options" to beautify the function signature: https://github.com/STRRL/rfcs/blob/logging/text/2021-12-09-logging.md#using-logger-as-function-parameter-is-ugly-is-there-another-way